libuavcan/src/transport/can_acceptance_filter_configurator.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>,
+ *                    Ilia  Sheremet <illia.sheremet@gmail.com>
+ */
+
+#include <uavcan/transport/can_acceptance_filter_configurator.hpp>
+#include <cassert>
+
+namespace uavcan
+{
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterMsgMask;
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterServiceRequestID;
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterServiceRequestMask;
+const unsigned CanAcceptanceFilterConfigurator::ServiceRespFrameID;
+const unsigned CanAcceptanceFilterConfigurator::ServiceRespFrameMask;
+
+int16_t CanAcceptanceFilterConfigurator::fillArray()
+{
+    multiset_configs_.clear();
+    uint16_t i = 0;
+
+    CanFilterConfig ServiseRespFrame;
+    ServiseRespFrame.id = ServiceRespFrameID;
+    ServiseRespFrame.mask = ServiceRespFrameMask;
+    multiset_configs_.emplace(ServiseRespFrame);
+    if (multiset_configs_.getByIndex(i) == NULL)
+    {
+        UAVCAN_ASSERT(0);
+    }
+    i++;
+
+    const TransferListenerBase* p = node_.getDispatcher().getListOfMessageListeners().get();
+    while (p)
+    {
+        CanFilterConfig cfg;
+        cfg.id = static_cast<uint32_t>(p->getDataTypeDescriptor().getID().get()) << 16;
+        cfg.id |= static_cast<uint32_t>(p->getDataTypeDescriptor().getKind()) << 8;
+        cfg.mask = DefaultFilterMsgMask;
+
+        multiset_configs_.emplace(cfg);
+        if (multiset_configs_.getByIndex(i) == NULL)
+        {
+            UAVCAN_ASSERT(0);
+        }
+
+        i++;
+        p = p->getNextListNode();
+    }
+
+    const TransferListenerBase* p1 = node_.getDispatcher().getListOfServiceRequestListeners().get();
+    while (p1)
+    {
+        CanFilterConfig cfg;
+        cfg.id = DefaultFilterServiceRequestID;
+        cfg.id |= static_cast<uint32_t>(p1->getDataTypeDescriptor().getID().get()) << 17;
+        cfg.mask = DefaultFilterServiceRequestMask;
+
+        multiset_configs_.emplace(cfg);
+        if (multiset_configs_.getByIndex(i) == NULL)
+        {
+            UAVCAN_ASSERT(0);
+        }
+
+        i++;
+        p1 = p1->getNextListNode();
+    }
+
+    if (multiset_configs_.getSize() == 0)
+    {
+        return -ErrDriver;
+    }
+
+    for (uint16_t i = 0; i < multiset_configs_.getSize(); i++)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::fillArray()", "map.ID [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->id);
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::fillArray()", "map.MK [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->mask);
+    }
+
+    return 0;
+}
+
+int16_t CanAcceptanceFilterConfigurator::computeConfiguration()
+{
+    uint16_t i_rank = 0, j_rank = 0;
```

Comment:
so it means that we could calculate configurations maximum for 256 listeners? Does it suit us?